File: 110.balanced-binary-tree.cpp
/*
 * @lc app=leetcode id=110 lang=cpp
 *
 * [110] Balanced Binary Tree
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, int &depth) {
        if(!root) return true;

        int leftDepth = 0;
        int rightDepth = 0;
        if(helper(root->left, leftDepth) && helper(root->right, rightDepth)) {
            ++depth;
            depth += max(leftDepth, rightDepth);
            return abs(leftDepth - rightDepth) < 2;
        }

        return false;
    }

    bool isBalanced(TreeNode* root) {
        if(!root) return true;
        int leftDepth = 0;
        int rightDepth = 0;
        if(helper(root->left, leftDepth) && helper(root->right, rightDepth)) {
            return abs(leftDepth - rightDepth) < 2;
        }
        return false;
    }
};

File: code/110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if self.result == False:
            return 0
        if root == None:
            return 0
        left_depth = self.maxDepth(root.left)
        right_depth = self.maxDepth(root.right)
        if abs(left_depth - right_depth) > 1:
            self.result = False
        return 1 + max(left_depth, right_depth)

    def isBalanced(self, root: TreeNode) -> bool:
        if root == None:
            return True
        self.result = True
        self.maxDepth(root)
        return self.result
# @lc code=end